tikv/raftstore/msg.go

```diff
@@ -0,0 +1,145 @@
+package raftstore
+
+import (
+	"time"
+
+	"github.com/coreos/etcd/raft"
+	"github.com/ngaut/unistore/rocksdb"
+	"github.com/pingcap/kvproto/pkg/import_sstpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raft_cmdpb"
+	"github.com/pingcap/kvproto/pkg/raft_serverpb"
+)
+
+type MsgType int
+
+const (
+	MsgTypeRaftMessage            MsgType = 1 //
+	MsgTypeRaftCmd                MsgType = 2
+	MsgTypeSplitRegion            MsgType = 3
+	MsgTypeComputeResult          MsgType = 4
+	MsgTypeRegionApproximateSize  MsgType = 5
+	MsgTypeRegionApproximateKeys  MsgType = 6
+	MsgTypeCompactionDeclineBytes MsgType = 7
+	MsgTypeHalfSplitRegion        MsgType = 8
+	MsgTypeMergeResult            MsgType = 9
+	MsgTypeGcSnap                 MsgType = 10
+	MsgTypeClearRegionSize        MsgType = 11
+	MsgTypeTick                   MsgType = 12
+	MsgTypeSignificantMsg         MsgType = 13
+	MsgTypeStart                  MsgType = 14
+	MsgTypeApplyRes               MsgType = 15
+	MsgTypeNoop                   MsgType = 16
+
+	MsgTypeStoreRaftMessage       MsgType = 101
+	MsgTypeStoreSnapshotStats     MsgType = 102
+	MsgTypeStoreValidateSSTResult MsgType = 103
+	// Clear region size and keys for all regions in the range, so we can force them to re-calculate
+	// their size later.
+	MsgTypeStoreClearRegionSizeInRange MsgType = 104
+	MsgTypeStoreCompactedEvent         MsgType = 105
+	MsgTypeStoreTick                   MsgType = 106
+	MsgTypeStoreStart                  MsgType = 107
+)
+
+type Callback func(resp *raft_cmdpb.RaftCmdResponse)
+
+type PeerTick int
+
+const (
+	PeerTickRaft             PeerTick = 1
+	PeerTickRaftLogGC        PeerTick = 2
+	PeerTickSplitRegionCheck PeerTick = 3
+	PeerTickPdHeartbeat      PeerTick = 4
+	PeerTickPeerStaleState   PeerTick = 5
+)
+
+type StoreTick int
+
+const (
+	StoreTickCompactCheck     StoreTick = 1
+	StoreTickPdStoreHeartbeat StoreTick = 2
+	StoreTickSnapGC           StoreTick = 3
+	StoreTickCompactLockCF    StoreTick = 4
+	StoreTickConsistencyCheck StoreTick = 5
+	StoreTickCleanupImportSSI StoreTick = 6
+)
+
+type SignificantMsg struct {
```

Comment:
I prefer to use a type field to indicate which message type this is, and there may be new messages later.
```
type SignificantMsg struct {
    MsgType int
    Unreachable *xxx
    SnapshotStatus *xxx
}
```
```
pub enum SignificantMsg {
    /// Reports whether the snapshot sending is successful or not.
    SnapshotStatus {
        region_id: u64,
        to_peer_id: u64,
        status: SnapshotStatus,
    },
    /// Reports `to_peer_id` is unreachable.
    Unreachable { region_id: u64, to_peer_id: u64 },
}
```